Spelling error decorations and suggestions (Partial for #388)

matthew-carroll opened the pull request:
Spelling error decorations and suggestions (Partial for #388)


https://github.com/user-attachments/assets/9157f26a-50b9-4af9-8309-1945a1728be7

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_and_grammar_plugin.dart:
```diff
@@ -0,0 +1,280 @@
+
+  final _styler = SpellingAndGrammarStyler();
+
+  final _selectedWordLink = LeaderLink();
```
Maybe we could add a dart doc for this.

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_and_grammar_plugin.dart:
```diff
@@ -0,0 +1,280 @@
+
+    for (final textNodeId in changedTextNodes) {
+      final textNode = document.getNodeById(textNodeId);
+      if (textNode == null) {
+        // TODO: log a warning.
+        continue;
+      }
+      if (textNode is! TextNode) {
+        // TODO: log a warning.
+        continue;
+      }
```
Should this logs be added?

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_and_grammar_plugin.dart:
```diff
@@ -0,0 +1,280 @@
+    // the response out of order with other requests.
+    _asyncRequestIds[textNode.id] ??= 0;
+    final requestId = _asyncRequestIds[textNode.id]! + 1;
+    _asyncRequestIds[textNode.id] = _asyncRequestIds[textNode.id]! + 1;
```
Can't we use `requestId` here?

matthew-carroll replied:
Use it where, exactly? Also, are you saying there's a problem with using the `textNode.id`? Or are you suggesting using the `requestId` for another reason?

angelosilvestre replied:
I mean to do:

```dart
_asyncRequestIds[textNode.id] = requestId;
```

instead of re-incrementing `_asyncRequestIds[textNode.id]`

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_and_grammar_plugin.dart:
```diff
@@ -0,0 +1,280 @@
+    _asyncRequestIds[textNode.id] = _asyncRequestIds[textNode.id]! + 1;
+
+    int startingOffset = 0;
+    TextRange prevError = TextRange.empty;
```
Wouldn't `lastError` be better than `prevError` here?

matthew-carroll replied:
Is there a reason that it would be? I think "last error" might actually be more confusing. From the perspective of the paragraph, the "last error" would be the final error that's discovered in the paragraph. At least with a term like "previous", we're implying a previous/next situation, which hopefully makes it clear that we're scanning through text. But you let me know what you think.

angelosilvestre replied:
I guess it makes sense.

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_and_grammar_plugin.dart:
```diff
@@ -0,0 +1,280 @@
+    final spellingSuggestions = <TextRange, SpellingErrorSuggestion>{};
+    if (isSpellCheckEnabled) {
+      do {
+        prevError = await spellChecker.checkSpelling(
```
Shouldn't we pass the language to `checkSpelling` and `guesses`?

matthew-carroll replied:
You tell me...should we? If so, can you please let me know the code I should use to obtain it here?

angelosilvestre replied:
I think we should pass the language to avoid any surprises, the code would be:

```dart
final locale = PlatformDispatcher.instance.locale;
final language = spellChecker.convertDartLocaleToMacLanguageCode(locale)!;
```

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,439 @@
+    setState(() {
+      // If the selection was sitting in an ignored spelling error, and
+      // now the selection is somewhere else, reset the ignored error.
+      if (_ignoredSpellingErrorRange != null) {
```
How about returning early if `_ignoredSpellingErrorRange` is `null`?

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,439 @@
+  ) {
+    _suggestionListenable.value = null;
+
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      if (_suggestionToolbarOverlayController.isShowing) {
+        _suggestionToolbarOverlayController.hide();
+      }
+    });
```
Can you add a comment explaining why we hide the overlay here?

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,439 @@
+      return const SizedBox();
+    }
+
+    return OverlayPortal(
```
Can you add a comment explaining why we need an `OverlayPortal`, since the widget returned here is already displayed over the content? I'm assuming it's because of the `Follower`, but it would be good to explain that here.

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestions.dart:
```diff
@@ -0,0 +1,120 @@
+    notifyListeners();
+  }
+
+  /// Clears all spelling suggestions for text within the node with the given [nodeId].
+  void clearNode(String nodeId) {
+    _suggestions.remove(nodeId);
+  }
+
+  /// Clears all spelling suggestions for all text in the document.
+  void clear() {
+    _suggestions.clear();
+  }
```
Since we are notifying listeners when adding suggestions, shouldn't we also notify when clearing?

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestions.dart:
```diff
@@ -0,0 +1,120 @@
+
+class SpellingErrorSuggestion {
+  const SpellingErrorSuggestion(
+      {required this.word, required this.nodeId, required this.range, required this.suggestions});
```
How about end with `,` to re-format?

miguelcmedeiros commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,439 @@
+      parentNode: widget.editorFocusNode,
+      child: Container(
+        decoration: BoxDecoration(
+          color: Colors.white,
```
The current implementation looks fine on light mode, but not on dark mode.
Also, it would be great to be able to customize the suggestion overlay. Is it possible to expose this in the API?

miguelcmedeiros reviewed (commented):
@matthew-carroll noticed a small issue (see comment) with the code.

Also tested the feature. Here's my initial feedback:
* We should be able to pass a `tapRegionGroupId` to the suggestion toolbar so that it plays well with a `TapRegion` ancestor
* Tested on android for potential regression and the following is thrown:
```
I/flutter ( 5003): (12.427) PlatformException(channel-error, Unable to establish connection on channel: "dev.flutter.pigeon.super_editor_spellcheck.SpellCheckMac.checkSpelling"., null, null)
I/flutter ( 5003): #0      SpellCheckMac.checkSpelling (package:super_editor_spellcheck/src/platform/messages.g.dart:277:7)
I/flutter ( 5003): <asynchronous suspension>
I/flutter ( 5003): #1      SuperEditorSpellCheckerMacPlugin.checkSpelling (package:super_editor_spellcheck/src/platform/spell_checker_mac.dart:37:20)
I/flutter ( 5003): <asynchronous suspension>
I/flutter ( 5003): #2      SpellingAndGrammarReaction._findSpellingAndGrammarErrors (package:super_editor_spellcheck/src/super_editor/spelling_and_grammar_plugin.dart:200:21)
I/flutter ( 5003): <asynchronous suspension>
```

matthew-carroll commented:
@miguelcmedeiros - I addressed your issues.


https://github.com/user-attachments/assets/f0733c52-8bfc-4008-a71f-19bf1ae1a2cb

Do you want to try out the configurable toolbar builder with tap region ID to make sure you can build the kind of toolbar you want with the given API?

The configuration point is at the plugin level. You pass your own toolbar builder to the plugin.

matthew-carroll commented:
Test failure is due to non-existent screenshot files that currently exist on `main`. We need to copy those to `stable`.

miguelcmedeiros commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,502 @@
+
+  /// Builder that creates the spelling suggestion toolbar, which appears near
+  /// the currently selected mis-spelled word.
+  final SpellingErrorSuggestionToolbarBuilder toolbarBuilder;
```
This builder is not being called.

miguelcmedeiros commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,502 @@
+typedef SpellingErrorSuggestionToolbarBuilder = Widget Function(
+  BuildContext context, {
+  required FocusNode editorFocusNode,
+  Object? tapRegionId,
```
`tapRegionId` probably doesn't need to be part of this typedef since `SpellingErrorSuggestionOverlay` does not know about this field.

miguelcmedeiros commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
This file should be exported so that it can be used from the caller side, e.g. when passing a `toolbarBuilder` to the plugin that returns `DesktopSpellingSuggestionToolbar` but with a caller-side value for `tapRegionId`

miguelcmedeiros commented:
> Do you want to try out the configurable toolbar builder with tap region ID to make sure you can build the kind of toolbar you want with the given API?

@matthew-carroll tried to give it another go with the latest changes, but it was not possible. Left a few comments related to why it was not possible.

matthew-carroll commented:
@miguelcmedeiros I updated the widget tree to use the toolbar builder, exported all source files, and removed the tap region ID from the toolbar build interface.

matthew-carroll commented:
@miguelcmedeiros any comments at all? any adjustments? do you think you should try to ship with this PR before we merge, or do you feel good about us merging this?

matthew-carroll commented on super_editor_spellcheck/.run/Mac Plugin Sample.run.xml:
Yeah - we always commit Android Studio run configurations. If you think one of them shouldn't exist, please let me know.

matthew-carroll commented:
@angelosilvestre can you take another look. I resolved most of your comments, and then followed up with a few of them.

I think as soon as you approve this PR, we're ready to merge in. Superlist seems happy with it.

angelosilvestre commented on super_editor_spellcheck/lib/src/super_editor/spelling_error_suggestion_overlay.dart:
```diff
@@ -0,0 +1,507 @@
+    _suggestionListenable.value = spellingSuggestion;
+
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      _suggestionToolbarOverlayController.show();
```
Should we check if we are still mounted?

angelosilvestre reviewed (approved):
LGTM with a couple comments. Also, I'm still seeing a bunch of golden files that previously were showing an underline, and now aren't. I wanted to make sure this is correct.

matthew-carroll commented:
@angelosilvestre do you wanna do a final pass? I think the goldens should be OK now, but let me know if you see anything wrong there.

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_android_no-stylesheet.png:
The underline was removed from this golden. Is that correct?

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_android_with-stylesheet.png:
Another golden with the underline removed.

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_iOS_no-stylesheet.png:
Another golden with the underline removed.

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_iOS_with-stylesheet.png:
Another golden with the underline removed.

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_macOS_no-stylesheet.png:
Another golden with the underline removed.

angelosilvestre commented on super_editor/test_goldens/editor/text_entry/goldens/super-editor_text-entry_spelling-error-shows-underline_task_macOS_with-stylesheet.png:
Another golden with the underline removed.

angelosilvestre reviewed (commented):
Code looks good. We still have some golden files that are removing the underlines. I left a comment at each of them.

matthew-carroll commented:
@angelosilvestre I fixed a task spelling error bug - can you check the goldens again?

angelosilvestre reviewed (approved):
LGTM
